File: packages/capacitor-dragonfly-sensor-bridge/ios/Plugin/DragonflySensorBridgePlugin.swift
import Capacitor
import Foundation
import Security

/// Capacitor plugin that satisfies the four-method BridgeAdapter
/// contract from `apps/patient-pwa/src/bridge/types.ts`.
///
/// Storage: the bearer token lives in the iOS Keychain
/// (`kSecClassGenericPassword`, `kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly`).
/// The WebView never sees the raw token after install — `getStatus()`
/// only reports `tokenInstalled: true/false`. The patient binding
/// (`patientId` the token was issued for) is stored alongside so
/// `emitDemoReading` can refuse cross-patient writes locally as well.
///
/// Network: the plugin owns the `POST /api/glucose/sync` request and
/// attaches the bearer itself. Reading is tagged honestly
/// (`vendor: "unknown"`, `deviceName: "Dragonfly Demo Bridge"`,
/// `source: "manual"`) — emitting synthetic data as if it were real
/// Dexcom or Libre output is a contract violation.
///
/// API base URL comes from the plugin config in `capacitor.config.ts`:
///   plugins: { DragonflySensorBridge: { apiBase: "https://..." } }
@objc(DragonflySensorBridgePlugin)
public class DragonflySensorBridgePlugin: CAPPlugin {

    private let service = "com.dragonfly.sensorbridge"
    private let tokenAccount = "bridgeToken"
    private let patientAccount = "bridgePatientId"
    private let lastSyncDefaultsKey = "DragonflySensorBridge.lastSyncedAt"

    @objc func installToken(_ call: CAPPluginCall) {
        guard let token = call.getString("token"), !token.isEmpty else {
            call.reject("token must be a non-empty string")
            return
        }
        do {
            try keychainSet(account: tokenAccount, value: token)
            call.resolve()
        } catch {
            call.reject("keychain write failed: \(error)")
        }
    }

    @objc func clearToken(_ call: CAPPluginCall) {
        keychainDelete(account: tokenAccount)
        keychainDelete(account: patientAccount)
        UserDefaults.standard.removeObject(forKey: lastSyncDefaultsKey)
        call.resolve()
    }

    @objc func getStatus(_ call: CAPPluginCall) {
        let installed = (try? keychainGet(account: tokenAccount)) != nil
        call.resolve([
            "tokenInstalled": installed,
            "adapterLabel": "iOS native bridge (Capacitor)",
            "lastSyncedAt": UserDefaults.standard.string(forKey: lastSyncDefaultsKey) as Any,
            "pendingCount": 0,
        ])
    }

    /// Libre NFC tap-reading is Android-only in V1.5. iOS implementation
    /// needs the `com.apple.developer.nfc.readersession.formats` entitlement
    /// and a different NFC API (Core NFC `NFCTagReaderSession`); deferred
    /// to V2+.
    @objc func readLibreOnce(_ call: CAPPluginCall) {
        call.unimplemented("readLibreOnce is Android-only in V1.5. iOS NFC support is V2+.")
    }

    /// Dexcom G6 / G7 brief BLE reads are Android-only in V1.5. iOS needs
    /// the `com.apple.developer.bluetooth-central-background` private
    /// entitlement to sustain BLE long enough to catch the next
    /// 5-minute G6 broadcast; deferred to V2+.
    @objc func readDexcomG6Once(_ call: CAPPluginCall) {
        call.unimplemented("readDexcomG6Once is Android-only in V1.5.")
    }

    @objc func readDexcomG7Once(_ call: CAPPluginCall) {
        call.unimplemented("readDexcomG7Once is Android-only in V1.5.")
    }

    @objc func emitDemoReading(_ call: CAPPluginCall) {
        guard let patientId = call.getString("patientId") else {
            call.reject("patientId required")
            return
        }
        guard let valueMgDl = call.getDouble("valueMgDl") else {
            call.reject("valueMgDl required")
            return
        }
        let token: String
        do {
            guard let stored = try keychainGet(account: tokenAccount) else {
                call.reject("no bridge token installed")
                return
            }
            token = stored
        } catch {
            call.reject("keychain read failed: \(error)")
            return
        }

        // Patient binding: if a patientId was previously associated with
        // this token, refuse cross-patient writes locally. The Worker
        // enforces this server-side too; this is the cheap first line.
        if let bound = try? keychainGet(account: patientAccount), bound != patientId {
            call.reject("token bound to a different patientId")
            return
        }
        if (try? keychainGet(account: patientAccount)) == nil {
            try? keychainSet(account: patientAccount, value: patientId)
        }

        guard let apiBase = (getConfig().getString("apiBase"))
            .flatMap({ URL(string: $0) }) else {
            call.reject("apiBase missing in capacitor.config.ts plugins.DragonflySensorBridge")
            return
        }

        Task {
            do {
                let result = try await postSync(
                    apiBase: apiBase,
                    token: token,
                    patientId: patientId,
                    valueMgDl: valueMgDl
                )
                UserDefaults.standard.set(result.timestamp, forKey: lastSyncDefaultsKey)
                call.resolve([
                    "accepted": result.accepted,
                    "duplicates": result.duplicates,
                    "rejected": result.rejected.map {
                        ["index": $0.index, "reason": $0.reason]
                    },
                ])
            } catch {
                call.reject("sync failed: \(error)")
            }
        }
    }

    // MARK: - Network

    private struct SyncReply {
        let accepted: Int
        let duplicates: Int
        let rejected: [Rejection]
        let timestamp: String
        struct Rejection { let index: Int; let reason: String }
    }

    private func postSync(
        apiBase: URL,
        token: String,
        patientId: String,
        valueMgDl: Double
    ) async throws -> SyncReply {
        var url = apiBase
        url.appendPathComponent("api")
        url.appendPathComponent("glucose")
        url.appendPathComponent("sync")

        let nowIso = ISO8601DateFormatter().string(from: Date())
        let body: [String: Any] = [
            "patientId": patientId,
            "vendor": "unknown",
            "deviceName": "Dragonfly Demo Bridge",
            "bridgeVersion": "ios-shell-0.1.0",
            "readings": [[
                "patientId": patientId,
                "valueMgDl": valueMgDl,
                "source": "manual",
                "vendor": "unknown",
                "deviceName": "Dragonfly Demo Bridge",
                "context": chooseContext(),
                "timestamp": nowIso,
                "readingKind": "manual",
                "ingestionPath": "manual",
                "rawDeviceId": "ios-shell-\(nowIso)",
                "notes": "Synthetic demo reading from the iOS Capacitor shell.",
            ]],
        ]

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw NSError(domain: "DragonflySensorBridge", code: -1)
        }
        guard (200..<300).contains(http.statusCode) else {
            let txt = String(data: data, encoding: .utf8) ?? ""
            throw NSError(
                domain: "DragonflySensorBridge",
                code: http.statusCode,
                userInfo: [NSLocalizedDescriptionKey: txt]
            )
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw NSError(domain: "DragonflySensorBridge", code: -2)
        }
        let rejected = (json["rejected"] as? [[String: Any]])?.compactMap { r -> SyncReply.Rejection? in
            guard let i = r["index"] as? Int, let reason = r["reason"] as? String else { return nil }
            return SyncReply.Rejection(index: i, reason: reason)
        } ?? []
        return SyncReply(
            accepted: json["accepted"] as? Int ?? 0,
            duplicates: json["duplicates"] as? Int ?? 0,
            rejected: rejected,
            timestamp: nowIso
        )
    }

    private func chooseContext() -> String {
        let hour = Calendar.current.component(.hour, from: Date())
        switch hour {
        case ..<9:  return "pre_taiyi"
        case ..<11: return "post_taiyi"
        case ..<13: return "before_lunch"
        case ..<16: return "post_lunch_1_to_2h"
        case ..<19: return "post_lunch_3_to_4h"
        default:    return "end_of_day"
        }
    }

    // MARK: - Keychain

    private func keychainSet(account: String, value: String) throws {
        let data = value.data(using: .utf8) ?? Data()
        let baseQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(baseQuery as CFDictionary)
        var add = baseQuery
        add[kSecValueData as String] = data
        add[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        let status = SecItemAdd(add as CFDictionary, nil)
        guard status == errSecSuccess else {
            throw NSError(domain: "Keychain", code: Int(status))
        }
    }

    private func keychainGet(account: String) throws -> String? {
        let q: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(q as CFDictionary, &item)
        if status == errSecItemNotFound { return nil }
        guard status == errSecSuccess, let data = item as? Data else {
            throw NSError(domain: "Keychain", code: Int(status))
        }
        return String(data: data, encoding: .utf8)
    }

    private func keychainDelete(account: String) {
        let q: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(q as CFDictionary)
    }
}

File: capacitor-dragonfly-sensor-bridge/android/src/main/java/com/dragonfly/shell/sensorbridge/DragonflySensorBridgePlugin.kt
package com.dragonfly.shell.sensorbridge

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * Capacitor plugin satisfying the four-method BridgeAdapter contract
 * from `apps/patient-pwa/src/bridge/types.ts`.
 *
 * Storage: bearer token + bound patient id live in
 * EncryptedSharedPreferences (AES-256-GCM with HKDF-4KB streaming, master
 * key in the Android Keystore). The WebView never sees the raw token
 * after install — `getStatus()` only reports `tokenInstalled: true/false`.
 *
 * Network: this plugin owns the `POST /api/glucose/sync` request and
 * attaches the bearer itself. The reading is tagged honestly
 * (`vendor: "unknown"`, `deviceName: "Dragonfly Demo Bridge"`,
 * `source: "manual"`).
 *
 * API base URL comes from the plugin config in `capacitor.config.ts`:
 *   plugins: { DragonflySensorBridge: { apiBase: "https://..." } }
 */
@CapacitorPlugin(name = "DragonflySensorBridge")
class DragonflySensorBridgePlugin : Plugin() {

    private val prefsName = "dragonfly_sensor_bridge"
    private val keyToken = "bridgeToken"
    private val keyPatient = "bridgePatientId"
    private val keyLastSync = "lastSyncedAt"
    private val scope = CoroutineScope(Dispatchers.IO)

    private fun prefs(): SharedPreferences {
        val ctx: Context = context.applicationContext
        val masterKey = MasterKey.Builder(ctx)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        return EncryptedSharedPreferences.create(
            ctx,
            prefsName,
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    }

    @PluginMethod
    fun installToken(call: PluginCall) {
        val token = call.getString("token")?.takeIf { it.isNotBlank() }
            ?: return call.reject("token must be a non-empty string")
        prefs().edit().putString(keyToken, token).apply()
        call.resolve()
    }

    @PluginMethod
    fun clearToken(call: PluginCall) {
        prefs().edit().remove(keyToken).remove(keyPatient).remove(keyLastSync).apply()
        call.resolve()
    }

    @PluginMethod
    fun getStatus(call: PluginCall) {
        val p = prefs()
        val res = JSObject()
        res.put("tokenInstalled", p.contains(keyToken))
        res.put("adapterLabel", "Android native bridge (Capacitor)")
        res.put("lastSyncedAt", p.getString(keyLastSync, null))
        res.put("pendingCount", 0)
        call.resolve(res)
    }

    @PluginMethod
    fun emitDemoReading(call: PluginCall) {
        val patientId = call.getString("patientId")
            ?: return call.reject("patientId required")
        val valueMgDl = call.getDouble("valueMgDl")
            ?: return call.reject("valueMgDl required")

        val p = prefs()
        val token = p.getString(keyToken, null)
            ?: return call.reject("no bridge token installed")
        if (!checkPatientBinding(p, patientId)) {
            return call.reject("token bound to a different patientId")
        }
        val apiBase = config.getString("apiBase")
            ?: return call.reject("apiBase missing in capacitor.config.ts plugins.DragonflySensorBridge")

        scope.launch {
            try {
                val (result, ts) = postSync(
                    apiBase = apiBase,
                    token = token,
                    patientId = patientId,
                    valueMgDl = valueMgDl,
                    source = "manual",
                    vendor = "unknown",
                    deviceName = "Dragonfly Demo Bridge",
                    readingKind = "manual",
                    ingestionPath = "manual",
                    rawDeviceId = "android-shell-${isoNow()}",
                    notes = "Synthetic demo reading from the Android Capacitor shell.",
                    bridgeVersion = "android-shell-0.1.0",
                )
                p.edit().putString(keyLastSync, ts).apply()
                call.resolve(result)
            } catch (e: Throwable) {
                call.reject("sync failed: ${e.message}")
            }
        }
    }

    /**
     * Dexcom G6 brief BLE read. User opens the bridge screen, taps
     * "Read Dexcom G6", we connect to the transmitter named after their
     * transmitterId, authenticate, wait for the next 5-minute glucose
     * broadcast, POST it, disconnect. See DexcomG6BleReader.kt for the
     * protocol-level honesty notes.
     */
    @PluginMethod
    fun readDexcomG6Once(call: PluginCall) {
        val patientId = call.getString("patientId")
            ?: return call.reject("patientId required")
        val transmitterId = call.getString("transmitterId")
            ?: return call.reject("transmitterId required (6-char G6 transmitter ID)")
        val timeoutMs = call.getLong("timeoutMs") ?: 360_000L

        val p = prefs()
        val token = p.getString(keyToken, null)
            ?: return call.reject("no bridge token installed")
        if (!checkPatientBinding(p, patientId)) {
            return call.reject("token bound to a different patientId")
        }
        val apiBase = config.getString("apiBase")
            ?: return call.reject("apiBase missing in capacitor.config.ts plugins.DragonflySensorBridge")

        scope.launch {
            val read = runCatching {
                DexcomG6BleReader(context).readOnce(transmitterId, timeoutMs)
            }.getOrElse { return@launch call.reject("Dexcom G6 read failed: ${it.message}") }

            val out = JSObject()
                .put("vendor", "dexcom-g6")
                .put("transmitterIdLast2", read.transmitterIdLast2)
                .put("deviceName", read.deviceName)
                .put("valueMgDl", read.valueMgDl ?: JSONObject.NULL)
                .put("trend", read.trend)
                .put("timestampIso", read.timestampIso)
                .put("error", read.error)

            if (read.valueMgDl != null && read.timestampIso != null) {
                val (syncResult, ts) = runCatching {
                    postSync(
                        apiBase = apiBase,
                        token = token,
                        patientId = patientId,
                        valueMgDl = read.valueMgDl,
                        source = "cgm",
                        vendor = "dexcom",
                        deviceName = read.deviceName ?: "Dexcom G6",
                        readingKind = "sensor",
                        ingestionPath = "native-ble",
                        rawDeviceId = "g6-${read.transmitterIdLast2 ?: "unknown"}-${read.timestampIso}",
                        notes = null,
                        bridgeVersion = "android-ble-0.1.0",
                    )
                }.getOrElse {
                    out.put("syncError", it.message); call.resolve(out); return@launch
                }
                p.edit().putString(keyLastSync, ts).apply()
                out.put("sync", syncResult)
            }
            call.resolve(out)
        }
    }

    /**
     * Dexcom G7 brief BLE read. **Today this only gets as far as the
     * pre-J-PAKE handshake.** See DexcomG7BleReader.kt for the cert
     * problem and what shipping the full path needs.
     */
    @PluginMethod
    fun readDexcomG7Once(call: PluginCall) {
        val patientId = call.getString("patientId")
            ?: return call.reject("patientId required")
        val transmitterIdSuffix = call.getString("transmitterIdSuffix")
            ?: return call.reject("transmitterIdSuffix required (last 4 hex chars of G7 advert name)")
        val certJwt = call.getString("certJwt")
        val timeoutMs = call.getLong("timeoutMs") ?: 120_000L

        val p = prefs()
        val token = p.getString(keyToken, null)
            ?: return call.reject("no bridge token installed")
        if (!checkPatientBinding(p, patientId)) {
            return call.reject("token bound to a different patientId")
        }
        val apiBase = config.getString("apiBase")
            ?: return call.reject("apiBase missing in capacitor.config.ts plugins.DragonflySensorBridge")

        scope.launch {
            val read = runCatching {
                DexcomG7BleReader(context).readOnce(transmitterIdSuffix, certJwt, timeoutMs)
            }.getOrElse { return@launch call.reject("Dexcom G7 read failed: ${it.message}") }

            val out = JSObject()
                .put("vendor", "dexcom-g7")
                .put("deviceName", read.deviceName)
                .put("stage", read.stage.name)
                .put("valueMgDl", read.valueMgDl ?: JSONObject.NULL)
                .put("trend", read.trend)
                .put("timestampIso", read.timestampIso)
                .put("message", read.message)
                .put("error", read.error)

            if (read.valueMgDl != null && read.timestampIso != null) {
                val (syncResult, ts) = runCatching {
                    postSync(
                        apiBase = apiBase,
                        token = token,
                        patientId = patientId,
                        valueMgDl = read.valueMgDl,
                        source = "cgm",
                        vendor = "dexcom",
                        deviceName = read.deviceName ?: "Dexcom G7",
                        readingKind = "sensor",
                        ingestionPath = "native-ble",
                        rawDeviceId = "g7-${transmitterIdSuffix}-${read.timestampIso}",
                        notes = null,
                        bridgeVersion = "android-ble-0.1.0",
                    )
                }.getOrElse {
                    out.put("syncError", it.message); call.resolve(out); return@launch
                }
                p.edit().putString(keyLastSync, ts).apply()
                out.put("sync", syncResult)
            }
            call.resolve(out)
        }
    }

    /**
     * Libre NFC tap-read. Scope per docs/V1_5_PLAN.md: user-initiated,
     * one reading per tap. Libre 1 implemented; Libre 2 Gen1 detected and
     * FRAM captured but decryption unverified; Libre 2 Gen2 / Libre 3 /
     * Libre 3 Plus detected and reported unsupported.
     *
     * UNTESTED on real hardware in this commit — see the file-level
     * comment in LibreNfcReader.kt.
     */
    @PluginMethod
    fun readLibreOnce(call: PluginCall) {
        val patientId = call.getString("patientId")
            ?: return call.reject("patientId required")
        val timeoutMs = call.getLong("timeoutMs") ?: 30_000L

        val activity = activity
            ?: return call.reject("no host activity (plugin not attached)")
        val p = prefs()
        val token = p.getString(keyToken, null)
            ?: return call.reject("no bridge token installed")
        if (!checkPatientBinding(p, patientId)) {
            return call.reject("token bound to a different patientId")
        }
        val apiBase = config.getString("apiBase")
            ?: return call.reject("apiBase missing in capacitor.config.ts plugins.DragonflySensorBridge")

        scope.launch {
            val read = runCatching { LibreNfcReader.readOnce(activity, timeoutMs) }
                .getOrElse {
                    return@launch call.reject("NFC read failed: ${it.message}")
                }
            val out = JSObject()
                .put("sensorType", read.sensorType?.name)
                .put("sensorUid", read.sensorUid)
                .put("valueMgDl", read.valueMgDl ?: JSONObject.NULL)
                .put("timestampIso", read.timestampIso)
                .put("unsupportedReason", read.unsupportedReason)
                .put("error", read.error)

            // POST only when we actually extracted a usable reading.
            if (read.valueMgDl != null && read.timestampIso != null) {
                val (syncResult, ts) = runCatching {
                    postSync(
                        apiBase = apiBase,
                        token = token,
                        patientId = patientId,
                        valueMgDl = read.valueMgDl,
                        source = "cgm",
                        vendor = "libre",
                        deviceName = read.sensorType?.name?.lowercase()?.replace('_', ' ')
                            ?: "Libre",
                        readingKind = "sensor",
                        ingestionPath = "native-nfc",
                        rawDeviceId = "libre-${read.sensorUid ?: "unknown"}-${read.timestampIso}",
                        notes = null,
                        bridgeVersion = "android-nfc-0.1.0",
                    )
                }.getOrElse {
                    out.put("syncError", it.message)
                    call.resolve(out)
                    return@launch
                }
                p.edit().putString(keyLastSync, ts).apply()
                out.put("sync", syncResult)
            }
            call.resolve(out)
        }
    }

    private fun checkPatientBinding(p: SharedPreferences, patientId: String): Boolean {
        val bound = p.getString(keyPatient, null)
        if (bound != null && bound != patientId) return false
        if (bound == null) p.edit().putString(keyPatient, patientId).apply()
        return true
    }

    private fun postSync(
        apiBase: String,
        token: String,
        patientId: String,
        valueMgDl: Double,
        source: String,
        vendor: String,
        deviceName: String,
        readingKind: String,
        ingestionPath: String,
        rawDeviceId: String,
        notes: String?,
        bridgeVersion: String,
    ): Pair<JSObject, String> {
        val url = URL("${apiBase.trimEnd('/')}/api/glucose/sync")
        val nowIso = isoNow()

        val reading = JSONObject()
            .put("patientId", patientId)
            .put("valueMgDl", valueMgDl)
            .put("source", source)
            .put("vendor", vendor)
            .put("deviceName", deviceName)
            .put("context", chooseContext())
            .put("timestamp", nowIso)
            .put("readingKind", readingKind)
            .put("ingestionPath", ingestionPath)
            .put("rawDeviceId", rawDeviceId)
        if (notes != null) reading.put("notes", notes)

        val batch = JSONObject()
            .put("patientId", patientId)
            .put("vendor", vendor)
            .put("deviceName", deviceName)
            .put("bridgeVersion", bridgeVersion)
            .put("readings", JSONArray().put(reading))

        val conn = url.openConnection() as HttpURLConnection
        try {
            conn.requestMethod = "POST"
            conn.doOutput = true
            conn.setRequestProperty("Content-Type", "application/json")
            conn.setRequestProperty("Authorization", "Bearer $token")
            conn.outputStream.use { it.write(batch.toString().toByteArray(Charsets.UTF_8)) }
            val code = conn.responseCode
            val body = (if (code in 200..299) conn.inputStream else conn.errorStream)
                .bufferedReader().use { it.readText() }
            if (code !in 200..299) error("HTTP $code: $body")
            val json = JSONObject(body)
            val out = JSObject()
                .put("accepted", json.optInt("accepted"))
                .put("duplicates", json.optInt("duplicates"))
                .put("rejected", json.optJSONArray("rejected") ?: JSONArray())
            return out to nowIso
        } finally {
            conn.disconnect()
        }
    }

    private fun isoNow(): String {
        val fmt = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US)
        fmt.timeZone = TimeZone.getTimeZone("UTC")
        return fmt.format(Date())
    }

    private fun chooseContext(): String {
        val hour = Calendar.getInstance().get(Calendar.HOUR_OF_DAY)
        return when {
            hour < 9 -> "pre_taiyi"
            hour < 11 -> "post_taiyi"
            hour < 13 -> "before_lunch"
            hour < 16 -> "post_lunch_1_to_2h"
            hour < 19 -> "post_lunch_3_to_4h"
            else -> "end_of_day"
        }
    }
}
